### PR TITLE
Support providers in outgoing capabilities

### LINKED ISSUE #26163: Support providers to define capabilities of consumable configurations
### Expected Behavior

`Configuration.getOutgoing().capability(Object notation)` accepts providers and evaluates them late. Similar to how late attribute evaluation is supported (#18936).

For example, use cases like this are supported:

```
// early (e.g. in plugin)
configurations.create("somethingConsumable") {
  canBeResolved = false
  canBeConsumed = true
  ... // define attributes etc.
  outgoing.capability(provider { project.getGroup() + ":" + project.getName() + "-something:" + project.version })
}


// later (e.g. in build.gradle)
group = "org.example"
version = "1.1"
```

Concretely, I need this in a plugin:
https://github.com/gradlex-org/java-module-dependencies/blob/1f2dabe245a98e332d11e48436447828b0cf2da6/src/main/java/org/gradlex/javamodule/dependencies/JavaModuleVersionsPlugin.java#L64

### Current Behavior (optional)

_No response_

### Context

In previous Gradle versions, `ConfigurationInternal.beforeLocking { }` could be used as a place to configure Configuration details such as capabilities late (#19131). This method has been removed in Gradle 8.3 (#25126).

### PR DESCRIPTION
**Pull request by octylFractal**
Fixes #26163 

I think this needs more tests, but I'm not sure which ones would be good to duplicate for this.

### CODE REVIEW

File: platforms/software/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishJavaIntegTest.groovy [hunk 1]
```diff
@@ -958,6 +958,46 @@ class IvyPublishJavaIntegTest extends AbstractIvyPublishIntegTest {
         }
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/26163")
+    def "can publish java-library with lazy capabilities"() {
```
Comment (jvandort) [inline platforms/software/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishJavaIntegTest.groovy:962]: In addition to testing this through publishing, we should also verify that the capabilities are correct when tested through a `ResolutionResult`.

Also I find it odd that we have this under an "ivy" test. This really doesn't test any ivy behavior. Do we have tests more specific to GMM?

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java [hunk 2]
```diff
@@ -211,11 +211,15 @@ public void variants(Action<? super NamedDomainObjectContainer<ConfigurationVari
     @Override
     public void capability(Object notation) {
         if (canCreate) {
-            Capability descriptor = capabilityNotationParser.parseNotation(notation);
             if (capabilities == null) {
-                capabilities = Lists.newArrayListWithExpectedSize(1); // it's rare that a component would declare more than 1 capability
+                capabilities = domainObjectCollectionFactory.newDomainObjectCollection(Capability.class);
+            }
+            if (notation instanceof Provider) {
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java:217]: What are your thoughts on somehow moving this logic into the CapabilityNotationParserFactory? Right now, if I attempt to parse an incompatible type (for example pass an `Object` into this method), I get the following error:

```
Cannot convert the provided notation to an object of type Capability: java.lang.Object@6ccac6f4.
The following types/formats are supported:
  - Instances of Capability.
  - Instances of CharSequence.
  - Maps, for example [group: 'org.group', name: 'capability', version: '1.0'].
```

This message does not advertise that we support providers.

**Review by jvandort — COMMENTED** [review]
What are

File: platforms/software/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishJavaIntegTest.groovy (hunk 1 above)
Comment (octylFractal) [reply platforms/software/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishJavaIntegTest.groovy:962]: I only found tests related to Ivy and Maven specifically. This Ivy one seemed like the simplest one, I couldn't find an equivalent Maven test.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java (hunk 2 above)
Comment (octylFractal) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java:217]: I think we just have to live with that for now, perhaps we could add some flag in `NotationParser` to indicate that providers are mapped using it at a higher level, or we'd have to make the `NotationParser` produce a `Provider`, which would be more involved and possibly have issues with generics.

Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java:217]: If we don't add native support to the notation parser, we will also likely need to update the other places the parser is used.

Most importantly:
[DefaultMutableModuleDependencyCapabilitiesHandler](https://github.com/gradle/gradle/blob/1b03de5223aaea17cf044b88aa64af3c21a43d9b/subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultMutableModuleDependencyCapabilitiesHandler.java#L35-L37)

Maybe:
[DefaultCapabilitiesResolution](https://github.com/gradle/gradle/blob/1b03de5223aaea17cf044b88aa64af3c21a43d9b/platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolutionstrategy/DefaultCapabilitiesResolution.java#L155)

Comment (octylFractal) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java:217]: Done.

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/AbstractModuleDependency.java [hunk 3]
```diff
@@ -48,6 +50,7 @@ public abstract class AbstractModuleDependency extends AbstractDependency implem
 
     private ImmutableAttributesFactory attributesFactory;
     private NotationParser<Object, Capability> capabilityNotationParser;
+    private ObjectFactory objectFactory;
```
Comment (jvandort) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/AbstractModuleDependency.java:53]: One of my main worries here is that we're adding a new field to every dependency in Gradle. How does this affect memory usage for very large builds? I tried running a perf test but I think we need to rebase this PR on master first. 

In any case, I think in practice most of these dependencies _already_ have a reference to an object factory, given that they are almost always instantiated with an ObjectFactory already (and thus have an implicit generated ObjectFactory field). 

If we determine that this does not affect memory-usage/performance we are probably fine but I think we should test this

**Comment by octylFractal** [comment]
@bot-gradle test apt

File: platforms/software/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishJavaIntegTest.groovy (hunk 1 above)
Comment (jvandort) [reply platforms/software/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishJavaIntegTest.groovy:962]: I spoke with sterling about the testing here. He mentioned the possibility of dogfooding this feature by replacing the existing `ProjectDerivedCapability`. Do you think that is feasible here?

I still think we should have a `ResolutionResult` test in addition to a publishing one. 

In terms of maven, there is a similar `MavenPublishJavaIntegTest`, though in reality I feel that this feature is not java-related at all. We should be able to test this pretty easily without applying the java-library plugin at all.

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/AbstractModuleDependency.java (hunk 3 above)
Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/AbstractModuleDependency.java:53]: Perf tests passing so I guess this is fine. I think we can make improvements to the way we instantiate dependencies in general, though that's for another PR

File: platforms/software/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishJavaIntegTest.groovy (hunk 1 above)
Comment (jvandort) [reply platforms/software/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishJavaIntegTest.groovy:962]: I pushed up a commit with a new test and fixed a bug when using `requireCapability` with a provider. I think we can probably hold off on the dogfooding for another PR.

File: subprojects/core/src/main/java/org/gradle/api/internal/collections/DomainObjectCollectionFactory.java [hunk 4]
```diff
@@ -49,6 +49,11 @@ public interface DomainObjectCollectionFactory {
      */
     <T> NamedDomainObjectContainer<T> newNamedDomainObjectContainer(Class<T> type, Closure factoryClosure);
 
+    /**
+     * Creates a {@link DomainObjectCollection} for managing objects of the specified type.
+     */
+    <T> DomainObjectCollection<T> newDomainObjectCollection(Class<T> elementType);
```
Comment (jvandort) [inline subprojects/core/src/main/java/org/gradle/api/internal/collections/DomainObjectCollectionFactory.java:55]: As discussed, lets revert this and use a set to track capabilities.

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/ModuleDependencyCapabilitiesInternal.java [hunk 5]
```diff
@@ -17,11 +17,10 @@
 
 import org.gradle.api.artifacts.ModuleDependencyCapabilitiesHandler;
 import org.gradle.api.capabilities.Capability;
-
-import java.util.List;
+import org.gradle.api.provider.SetProperty;
 
 public interface ModuleDependencyCapabilitiesInternal extends ModuleDependencyCapabilitiesHandler {
-    List<Capability> getRequestedCapabilities();
+    SetProperty<Capability> getRequestedCapabilities();
```
Comment (jvandort) [inline subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/ModuleDependencyCapabilitiesInternal.java:23]: Can we expose this as a `SetProvider`? This interface is not necessarily mutable, only the implementation specifies `Mutable` in the name, and all external callers of this method only read the value.

**Review by jvandort — COMMENTED** [review]
LGTM otherwise

File: subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/ModuleDependencyCapabilitiesInternal.java (hunk 5 above)
Comment (octylFractal) [reply subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/ModuleDependencyCapabilitiesInternal.java:23]: I think you mean `Provider<Set<...>>`, but yes.

File: subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java [hunk 6]
```diff
@@ -49,7 +49,7 @@ public class DefaultDomainObjectCollection<T> extends AbstractCollection<T> impl
     private final CollectionEventRegister<T> eventRegister;
     private final ElementSource<T> store;
 
-    protected DefaultDomainObjectCollection(Class<? extends T> type, ElementSource<T> store, CollectionCallbackActionDecorator callbackActionDecorator) {
+    public DefaultDomainObjectCollection(Class<? extends T> type, ElementSource<T> store, CollectionCallbackActionDecorator callbackActionDecorator) {
```
Comment (jvandort) [inline subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java:52]: I think this can be made protected again.

**Comment by octylFractal** [comment]
@bot-gradle test this

**Review by jvandort — APPROVED** [review]

**Comment by octylFractal** [comment]
@bot-gradle test and merge

**Comment by octylFractal** [comment]
@bot-gradle cancel

**Comment by octylFractal** [comment]
@bot-gradle test and merge
